### ISSUE TITLE
Why do best practices suggest not updating setup.py?

### ISSUE THREAD

**Issue by adborden** [ckan/ckan#4590]
This is more of a question than an issue, hope you don't mind me asking here. I'm curious why the CKAN docs for writing extensions mention [using requirements.txt and **not** setup.py](https://github.com/ckan/ckan/blob/b9e45e2723d4abd70fa72b16ec4a0bebc795c56b/doc/extensions/best-practices.rst#add-third-party-libraries-to-requirementstxt)?
> If your extension requires third party libraries, rather than adding them to setup.py, they should be added to requirements.txt

I understand the idea of pinning dependencies, and requirements.txt makes a lot of sense for development, but why not also update setup.py with required dependencies and compatible version ranges? Without updating setup.py, it makes it much more difficult to manage dependency versions across extensions. The docs even call this out as a caveat to this approach:

> On the flip side, be mindful that this could also create version conflicts with requirements of considerably newer or older extensions.

### CKAN Version if known (or site URL)
2.8.2

### Please describe the expected behaviour
I can package an extension for pypi and have pip pull in the correct dependencies.

### Please describe the actual behaviour
Even if I install an extension from pypi, I have to install it in editable mode (`pip install -e`)  and install the extensions dependencies from requirements.txt.

### What steps can be taken to reproduce the issue? 
N/A

**Comment by wardi** [comment]
`install_requires` in setup.py is convenient for end users, it makes sure to grab the right versions of dependencies when using pip or setup.py install.

At the same time `install_requires` is very painful for developers because it prevents us from even running our code if *any* package is installed with *any* version conflict, even if we're just testing that a newer dependency works properly. Also many extensions support a wide range of ckan versions so choosing compatible ranges of dependencies is highly error-prone.

**Comment by adborden** [comment]
I agree, it's definitely more effort to keep setup.py up to date than not using it. But does not using setup.py really solve those problems of choosing compatible ranges? I think we're just working around it and pushing that work to the end user. If we encouraged the use of setup.py, it would distribute the work across extension maintainers rather than putting **all** the burden on end users.

Some extensions don't even document what versions of ckan or peer extensions they are compatible with (via README or docs) so as an end user you're stuck with guessing, with very little knowledge about which features extensions are using in dependent packages. Maintainers are in a much better position to figure out if an updated dependency is a problem or not. If there were version ranges in setup.py, you'd at least know there was an problem and have an idea of which extension you can fix it or report a bug.

I don't mean to be rude, but it sounds like you're saying `install_requires` would ensure we pull in the correct dependencies but because there is effort involved there we're okay with broken dependencies. Are there tricks or tools the community is using to ensure they don't end up with version conflicts and broken ckan instances?

With the current state of things, I have to do some of the leg work of figuring out the compatible version ranges for extensions I use anyway, so I would be happy to contribute that effort into moving towards setup.py support.

**Comment by davidread** [comment]
@adborden It's a reasonable point. I don't think this policy is a hard rule. I think one or two well-known extensions do specify ranges in setup.py's install_requires. I don't see too much wrong with that, when most of these extensions have no restrictions in versions (or at least very wide ranges). 

I guess there could be problems where the restrictions result in a conflict between two extensions, but one of the restrictions are due to a feature that is not needed for this site. Then all you can really do is fork.

But as I said I can't see problems in most cases, when the ranges are wide. The extension author should go the extra mile to understand what the minimum versions are. And really there should be no maximum versions, in this age of security and needing to keep up to date. So I'd support a change on this basis.

**Comment by davidread** [comment]
@adborden The need to add `-e` in the `pip install -e ckan` command I believe is just due to us never getting round to writing a complete manifest file, so some files don't get installed when you do `pip install ckan` currently. Having to remember the `-e` is a barrier to newcomers. I see there is [an attempt to fix it for core ckan](https://github.com/ckan/ckan/issues/3961) but it remains in the [docs](https://docs.ckan.org/en/latest/maintaining/installing/install-from-source.html) so maybe that can be done. But the extensions could also benefit from being installable without `-e` - they probably do, but it needs testing. If anyone wants to contribute at PR...
